crates/pathfinder/src/state/sync/l2.rs

```diff
@@ -441,58 +465,20 @@ pub async fn download_new_classes(
     .context("Joining database task")?
     .context("Querying database for missing classes")?;
 
-    for class_hash in require_downloading {
-        let class = download_class(sequencer, class_hash)
-            .await
-            .with_context(|| format!("Downloading class {}", class_hash.0))?;
-
-        match class {
-            DownloadedClass::Cairo { definition, hash } => tx_event
-                .send(SyncEvent::CairoClass { definition, hash })
+    let futures = require_downloading.into_iter().map(|class_hash| {
+        async move {
+            download_class(sequencer, class_hash)
                 .await
-                .with_context(|| {
-                    format!(
-                        "Sending Event::NewCairoContract for declared class {}",
-                        class_hash.0
-                    )
-                })?,
-            DownloadedClass::Sierra {
-                sierra_definition,
-                sierra_hash,
-                casm_definition,
-            } => {
-                // NOTE: we _have_ to use the same compiled_class_class hash as returned by the
-                // feeder gateway, since that's what has been added to the class
-                // commitment tree.
-                let Some(casm_hash) = state_update
-                    .declared_sierra_classes
-                    .iter()
-                    .find_map(|(sierra, casm)| (sierra.0 == class_hash.0).then_some(*casm))
-                else {
-                    // This can occur if the sierra was in here as a deploy contract, if the class
-                    // was declared in a previous block but not yet persisted by
-                    // the database.
-                    continue;
-                };
-                tx_event
-                    .send(SyncEvent::SierraClass {
-                        sierra_definition,
-                        sierra_hash,
-                        casm_definition,
-                        casm_hash,
-                    })
-                    .await
-                    .with_context(|| {
-                        format!(
-                            "Sending Event::NewSierraContract for declared class {}",
-                            class_hash.0
-                        )
-                    })?
-            }
+                .with_context(|| format!("Downloading class {}", class_hash.0))
         }
-    }
+        .in_current_span()
+    });
 
-    Ok(())
+    let stream = futures::stream::iter(futures).buffer_unordered(8);
```

Comment:
Maybe we'd like to make this configurable too?

Reply:
Good catch. That's the _only_ thing I wanted to make configurable and have apparently failed ;)

Reply:
I'd rather just leave this hard-coded to a smaller value (N) for now. `download_new_classes()` is being called from the future we're running concurrently we can have at most `fetch-concurrency * N` class downloads running. What's worse is that each of the downloads might actually invoke the compiler too (for Sierra classes) so we should probably move compilation to the rayon thread pool so that at least we have a sane upper limit for the number of compilers running in parallel.

Reply:
I've moved the compilation to a Rayon task in https://github.com/eqlabs/pathfinder/pull/2175/commits/d7c38b3da714a97f0d9b50408b5d027efbf52bca.